compute_mutation_probabilities.py

```diff
@@ -2,97 +2,140 @@
-    
 def run( args ):
-    # Load the mutation data
+    # Do some additional argument checking
+    if not args.weights_file and not args.permutation_directory:
+        sys.stderr.write('You must set the weights file or permutation directory, '\
+                         'otherwise nothing will be output.')
+        sys.exit(1)
+
+    # Load mutation data
     if args.verbose > 0:
         print '* Loading mutation data...'
-        
+
     mutation_data = load_mutation_data( args.mutation_file )
     genes, all_genes, patients, geneToCases, patientToMutations, params, hypermutators = mutation_data
-    num_all_genes, num_genes, num_patients = len(all_genes), len(genes), len(patients)
-    geneToIndex = dict(zip(all_genes, range(num_all_genes)))
-    patientToIndex = dict(zip(patients, range(num_patients)))
+
     geneToObserved = dict( (g, len(cases)) for g, cases in geneToCases.iteritems() )
     patientToObserved = dict( (p, len(muts)) for p, muts in patientToMutations.iteritems() )
-    
+    geneToIndex = dict( (g, i+1) for i, g in enumerate(all_genes) )
+    indexToGene = dict( (i+1, g) for i, g in enumerate(all_genes) )
+    patientToIndex = dict( (p, j+1) for j, p in enumerate(patients) )
+    indexToPatient = dict( (j+1, p) for j, p in enumerate(patients) )
+
+    edges = set()
+    for gene, cases in geneToCases.iteritems():
+        for patient in cases:
+            edges.add( (geneToIndex[gene], patientToIndex[patient]) )
+
+    edge_list = np.array(sorted(edges), dtype=np.int)
+
+    # Run the bipartite edge swaps
     if args.verbose > 0:
-        print '\tGenes:', num_all_genes
-        print '\tPatients:', num_patients
-        print '\tGenes mutated in at least one patient: {}'.format(num_genes)
-        
-    # Get the list of files we want to consider
-    permuted_files = args.permuted_files[:args.num_permutations]
-    print '* Analyzing {} permuted files...'.format(len(permuted_files))
-                                                    
-    # Set up the jobs
+        print '* Permuting matrices...'
+
+    m = len(all_genes)
+    n = len(patients)
+    num_edges = len(edges)
+    max_swaps = int(args.swap_multiplier*num_edges)
+    max_tries = 10**9
+    seeds = [ i+args.start_index for i in range(args.num_permutations) ]
+
+    # Run the bipartite edge swaps in parallel if more than one core indicated
     num_cores = args.num_cores if args.num_cores != -1 else mp.cpu_count()
     if num_cores != 1:
         pool = mp.Pool(num_cores)
         map_fn = pool.map
     else:
         map_fn = map
 
-    compute_args = [ (permuted_files[i::args.num_cores], geneToIndex, patientToIndex)
-                     for i in range(args.num_cores) ]
-
-    observeds = map_fn(compute_observed_wrapper, compute_args)
+    wrapper_args = [ (edge_list, max_swaps, max_tries, seeds[i::10*num_cores], 0, m,
+                      n, num_edges, indexToGene, indexToPatient) for i in range(10*num_cores) ]
```

Comment:
We should not bunch our runs into `num_cores` chunks.  The `chunksize` argument for `map` is a better way to do this, but not a priority.